Pal/src/host/Linux-SGX/enclave_framework.c

```diff
@@ -244,20 +244,49 @@ int load_trusted_file (PAL_HANDLE file, sgx_stub_t ** stubptr,
         goto failed;
 
     for (; offset < tf->size ; offset += TRUSTED_STUB_SIZE, s++) {
-        uint64_t mapping_size = tf->size - offset;
-        if (mapping_size > TRUSTED_STUB_SIZE)
-            mapping_size = TRUSTED_STUB_SIZE;
+        uint64_t mapping_size = MIN(tf->size - offset, TRUSTED_STUB_SIZE);
+        LIB_AESCMAC_CONTEXT aes_cmac;
+        ret = lib_AESCMACInit(&aes_cmac, (uint8_t *) &enclave_key,
+                              AES_CMAC_KEY_LEN);
+        if (ret < 0)
+            goto failed;
 
         ret = ocall_map_untrusted(fd, offset, mapping_size, PROT_READ, &umem);
         if (ret < 0)
             goto unmap;
 
-        lib_AESCMAC((void *) &enclave_key, AES_CMAC_KEY_LEN, umem,
-                    mapping_size, (uint8_t *) s, sizeof *s);
+        /*
+         * To prevent TOCTOU attack when generating the file checksum, we
+         * need to copy the file content into the enclave before hashing.
+         * For optimization, we use a relatively small buffer (1024 byte) to
+         * store the data for checksum generation.
+         */
+
+#define FILE_CHUNK_SIZE 1024
+
+        uint8_t small_chunk[FILE_CHUNK_SIZE]; /* Buffer for hashing */
+        int chunk_offset = 0;
+
+        for (; chunk_offset < mapping_size; chunk_offset += FILE_CHUNK_SIZE) {
+            uint64_t chunk_size = MIN(mapping_size - chunk_offset, FILE_CHUNK_SIZE);
+
+            /* Any file content needs to be copied into the enclave before
+             * checking and re-hashing */
+            memcpy(small_chunk, umem + chunk_offset, chunk_size);
+
+            /* Update the file checksum */
+            ret = lib_SHA256Update(&sha, small_chunk, chunk_size);
```

Comment:
Why are we doing both SHA and AES CMAC?

Reply:
For performance. Technically we can use any secure hash algorithm for the Merkel tree, but AES-CMAC is much faster and can use hardware acceleration.

Reply:
But why both?

Reply:
SHA256 is a more suitable algorithm for public-known secure digests, and is strong enough against brute-force. AES-CMAC needs a secret key unknown to the attackers, and such a secret key cannot be sent without a secured channel.

Overall the challenge here is that we need to officially define a hash algorithm as part of the manifest specification, but the hash algorithm used for the internal Merkel Trees can be anything. After experimenting with a few different algorithms, I found using AES-CMAC to be the fastest.

Reply:
I see.  I think the main remaining to-do is a bit of documentation here on how files are signed, and then we should be good to merge.